### PR TITLE
[v0.6] Populate backing namespace field for projects

### PR DESCRIPTION
**Pull request by JonCrowther**
## Issue: 
Builds on https://github.com/rancher/rancher/pull/47723

### CODE REVIEW

**Review by nflynt — DISMISSED** [review]

File: pkg/resources/management.cattle.io/v3/project/Project.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 ClusterName must be equal to the namespace, and must refer to an existing `management.cattle.io/v3.Cluster` object. In addition, users cannot update the field after creation.
 
+### BackingNamespace validation
+The BackingNamespace field cannot be changed once set. Projects without the BackingNamespace field can have it added.
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/Project.md:8]:
```suggestion
The `BackingNamespace` field cannot be changed once set. Projects without the `BackingNamespace` field can have it added.
```

File: pkg/resources/management.cattle.io/v3/project/Project.md [hunk 2]
```diff
@@ -31,4 +34,13 @@ If `field.cattle.io/no-creator-rbac` annotation is set, `field.cattle.io/creator
 
 ### On create
 
+Populates the BackingNamespace field by concatenating `Project.ClusterName` and `Project.Name`.
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/Project.md:37]:
```suggestion
Populates the `BackingNamespace` field by concatenating `Project.ClusterName` and `Project.Name`.
```

File: pkg/resources/management.cattle.io/v3/project/Project.md [hunk 3]
```diff
@@ -31,4 +34,13 @@ If `field.cattle.io/no-creator-rbac` annotation is set, `field.cattle.io/creator
...
+
+If the project is using a generated name (ie `GenerateName` is not empty), the generation happens within the mutating webhook.
+The reason for this is that the BackingNamespace is made up of the `Project.Name`, and name generation happens after mutating webhooks and before validating webhooks.
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/Project.md:40]:
```suggestion
The reason for this is that the `BackingNamespace` is made up of the `Project.Name`, and name generation happens after mutating and before validating webhooks.
```

File: pkg/resources/management.cattle.io/v3/project/Project.md [hunk 4]
```diff
@@ -31,4 +34,13 @@ If `field.cattle.io/no-creator-rbac` annotation is set, `field.cattle.io/creator
...
+
 Adds the authz.management.cattle.io/creator-role-bindings annotation.
+
+### On update
+
+If the BackingNamespace field is empty, populate the BackingNamespace field with the project name.
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/Project.md:46]:
```suggestion
If the `BackingNamespace` field is empty, it's populated with the project name.
```

File: pkg/resources/management.cattle.io/v3/project/mutator.go [hunk 5]
```diff
@@ -85,13 +96,81 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
...
+			// If the backing namespace already exists, generate a new project name
+			if err != nil && !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+		}
+	} else {
+		backingNamespace = name.SafeConcatName(newProject.Spec.ClusterName, strings.ToLower(newProject.Name))
+		_, err = m.namespaceClient.Get(backingNamespace, v1.GetOptions{})
+		if err == nil {
+			return nil, fmt.Errorf("failed to create project: namespace %v already exists", backingNamespace)
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/mutator.go:151]:
```suggestion
			return nil, fmt.Errorf("failed to create project: namespace %s already exists", backingNamespace)
```

Comment (pmatseykanets) [reply pkg/resources/management.cattle.io/v3/project/mutator.go:151]: Q: Should we throw an error or just make it a no-op? 🤔

File: pkg/resources/management.cattle.io/v3/project/mutator.go [hunk 6]
```diff
@@ -85,13 +96,81 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
...
+	}
+
+	newProject.Status.BackingNamespace = backingNamespace
+	return newProject, nil
 }
 
-func (m *Mutator) admitCreate(project *v3.Project, request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+// updateProjectNamespace fills in BackingNamespace with the project name if it wasn't already set
+// this was the naming convention of project namespaces prior to using the BackingNamespace field. Filling
+// it here is just to maintain backwards compatibility
```
Comment (pmatseykanets) [inline pkg/resources/management.cattle.io/v3/project/mutator.go:163]:
```suggestion
// updateProjectNamespace fills in BackingNamespace with the project name if it wasn't already set.
// This was the naming convention for project namespaces prior to using the BackingNamespace field.
// Filling it here is just to maintain backwards compatibility.
```

**Review by nflynt — APPROVED** [review]

**Review by pmatseykanets — APPROVED** [review]

**Pull request merged**